### DISCUSSION TITLE
Code is not scaling well

### DISCUSSION THREAD

**Discussion by tinusd** [casadi/casadi#3237] [Q&A] [answered]
Hello there

I have the following nonlinear optimization problem:
$\min f(x)$ 
$\text{ s.t. } g(x) = 0, \text{lb } \leq x \leq \text{ ub}$ 
where
$x \in \mathbb{R}^d$ and
$f(x) = \sum_{n \in \mathcal{N}} \Big(1/n{^2} {\sum}^d_{i=1}u(i)\cos(nx_i)\Big)^2$
$g(x) = m -  {\sum}^d_{i=1}u(i)\cos(x_i)$.
Here, $m$ is a scalar, $u$ is a vector of [1,-1,1,-1,...], and $\mathcal{N}$ is a set of harmonics; all of these are constant during optimization. 

I am busy comparing the performance of:
1. Computing the Jacobian and Hessian manually (I call Knitro directly from Matlab)
2. Using Casadi (I use nlpsol)
3. Using Casadi plus c-code generation (I use nlpsol).

I am using Matlab with Knitro as my solver. I give the same solver options to each case. I am running each problem 1000 times with the (same) randomly generated starting points. I use MX and I also enable extend. 

I am getting the following results (num_hars is the size of the set $\mathcal{N}$):

********* d = 10, num_hars = 66 *********
With analytical gradient and Hessian      : 17.07sec 
With casadi gradient and Hessian          : 10.33sec 
With casadi gradient and Hessian, code gen: 5.78sec

********* d = 20, num_hars = 66 *********
With analytical gradient and Hessian      : 31.39sec 
With casadi gradient and Hessian          : 25.46sec 
With casadi gradient and Hessian, code gen: 10.56sec 

********* d = 60, num_hars = 66 *********
With analytical gradient and Hessian      : 132.10sec 
With casadi gradient and Hessian          : 177.35sec 
With casadi gradient and Hessian, code gen: 88.15sec 

********* d = 60, num_hars = 2 *********
With analytical gradient and Hessian      : 30.31sec 
With casadi gradient and Hessian          : 11.66sec 
With casadi gradient and Hessian, code gen: 7.06sec 

It seems to be that, with how I coded, Casadi is not scaling as good as when I compute the gradient and Hessian manually. This seems to hold whether I scale the number of decision variables or the number of harmonics (the more harmonics, the better the approximation of my model). I attach the equations of my objective that I use for casadi:

```
function [f] = obj_cas(x,u,harms)

f=0;
for h=harms
    b = u'*cos(h*x); % inner product
    b = b/h^2;
    f = f + b^2;
end

```

Any idea what I could be doing suboptimally in the objective function? (Or could the problem be somewhere else?) I would greatly appreciate any input :)

**Comment by jaeandersson** [comment]
It's not clear to me that the CasADi code scales poorly, maybe CasADi has less overhead which is why it's faster for small problems? Maybe make a plot with more data points. If it's an actual issue and if you want to resolve this, you probably need to look into the sparsity pattern of the Hessian and see see if it's a matter of the graph coloring not working well for the particular structure you have.

**Reply by tinusd** [reply]
Thanks for your quick reply, Joel!

So, it could be case that CasADi has less overhead than my manually computed Jacobian and Hessian? But, then for large problems, the manually computed Jacobian and Hessian may be computionally more efficient? (BTW: I checked my Jacobian/Hessian against that of CasADi, and the error is within 1e-16).

For this particular problem, the Hessian is dense; there are no zeros. Do you have any other idea on what I should take a look at?

**Reply by jaeandersson** [reply]
CasADi communicates with solvers directly via C++. The fact that you're calling it from MATLAB is mostly irrelevant since all you do in MATLAB is to construct the problem, not do any actual evaluations. If your manual calculation of Hessians is in MATLAB, which means that every function evaluation includes evaluations in MATLAB, it might have a lot more overhead.

Would be hard for me to tell how to improve this without diving into the code.

**Reply by tinusd** [reply]
The manual calculation of the Jacobian and Hessian are in purely Matlab. Note: I do not use CasADi for this case: I call Knitro directly from Matlab.

I am sorry to be "the guy who posted his code", but if it is _immediately_ clear to you that I am doing something very wrong/suboptimal, or maybe not using CasADi correctly, it would be super helpful. 

The code is luckily short; I will only post the segment of the code where I use CasADi to construct the NLP (I won't post the section that relates to the manual computation of the Jacobian/Hessian). As I said, it seems for smaller problems, CasADi is doing very well and beats my manual Jacobian/Hessian. Then, when I increase the number of decision variables or the number of harmonics in the model, the tides turn and my manual Jacobian/Hessian starts winning.

This is the main:
```
%% problem settings
d = 60; % number of decision variables
harms = 3:2:60; % (odd) harmonics to consider; the more, the better the model approximation

u = ones(d,1); % create sequence of 1,-1,1,...
u(2:2:end) = -1;
m = 1; % equality constraint (constant) parameter

lb = 0*ones(d,1); % lower bound
ub = pi/2*ones(d,1); % upper bound

%% construct NLP
x = casadi.MX.sym('x',d);
nlp = struct('x',x, 'f',obj_cas(x,u,harms), 'g', nonlinconst_cas(x,u,m));

% casadi options
optis.expand = true;
optis.verbose = false;
optis.print_time = 0;
optis.options_file = 'knitro.opts';
optis.jit = false;
optis.compiler = 'shell';

solver = casadi.nlpsol('solver', 'knitro', nlp,optis);

%% solve and time problem
rng('default')
tic;
for i=1:100
x0 = rand(d,1)*pi/2;
 s = solver('x0',x0,...
 'lbg',0,'ubg',0, 'lbx',lb,'ubx',ub);
end
t2=toc;
```
and these are the objective and constraints:
```
function [f] = obj_cas(x,u,harms)
f=0;
for h=harms
    b = u'*cos(h*x); % inner product
    b = b/h^2;
    f = f + b^2;
end
```
```
function [c] = nonlinconst_cas(x,du,m)
c = m - 4/pi*du'*cos(x);
end
```

One thing: when I generate the C-code, the code is really, really long. I believe this is due to all the sums I am doing in my objective function. 

Thanks for the help, Joel. Much appreciated.

**Reply by jaeandersson** [reply, marked as answer]
I don't see anything obvious other than that I wouldn't do "expand = true". That could explain why your generated code gets long.

**Reply by tinusd** [reply]
OK, thanks a lot, Joel!
